src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+
+    These variables in BIOS RAM may be worth following:
+    - 0x20003ee0: FIQ enabled;
+    - 0x20003ee1: IRQ enabled;
+    Along with pairs used by interrupt callbacks:
+    - 0x20003ed1..0x20003ed4: status;
+    - 0x20003ee2..0x20003ee5: 1 if busy?
+*/
+TIMER_DEVICE_CALLBACK_MEMBER(sega_9h0_0008_state::scanline)
 {
-	map(0x00000000, 0x000001ff).rom().bankr("cartbank");
-	map(0x80000000, 0x807fffff).rom().bankr("cartbank");
+	irq_wait_speedup();
+
+	// Pixel clock approximation derived from palette lookup tables applied per scanline,
+	// used for gradient backgrounds in "Partner In TV".
+	//
+	// FIXME:
+	// - Rainbow arc in "Cars 2" before bridge section is not rendered;
+	// - Foreground drop animation in "Cooking Beena" title screen is skipped;
+	m_scanline = param / m_scale;
+	m_video_regs[0x4/4] = 0x10000 * (20 + m_scanline);
+
+	uint8_t irq_frequency = m_io_cpu_config->read() & 0xf;
+	if ((m_scanline % (SCREEN_H / irq_frequency)) == 0) {
+		uint8_t video_status = (m_video_regs[0] & 0x10) ^ 0x10;
+		m_video_regs[0] = (m_video_regs[0] & 0xffffffef) | video_status;
+
+		update_rtc();
+
+		request_irq();
+
+		// FIXME: Needed for .au data processing in "Kazoku Minna no Nouryoku Trainer"
+		m_requested_fiq = true;
+	}
+
+	request_fiq();
 }
 
+static const uint32_t irq_wait_signature[] = {
+   0xe3a03001,
+   0xe5c03000,
+   0xe5d03000,
+   0xe3530000,
+   0x1afffffc
+};
+
+void sega_9h0_0008_state::irq_wait_speedup()
+{
+	if ((m_io_cpu_config->read() & 0x10) == 0x10) {
+		return;
+	}
+
+	/*
+	    All games execute a busy wait until the next IRQ request is served.
+	    This can lead to significant downgrade of emulation speed.
+
+	    The busy wait subroutine is copied to a dynamic location in work RAM,
+	    somewhere after 0xc00cc000, but before the stack pointer. r0 stores
+	    an address to a variable that is updated by the game's IRQ callback
+	    when video data has been processed:
+
+	        e3 a0 30 01   mov     r3,#0x1
+	        e5 c0 30 00   strb    r3,[r0,#0x0]
+	        e5 d0 30 00   ldrb    r3,[r0,#0x0]
+	        e3 53 00 00   cmp     r3,#0x0
+	        1a ff ff fc   bne     LAB_c00ce8bc
+
+	    Epilogue is the following for most games:
+
+	        e5 9f 30 00   ldr     r3,[DAT_c00ce8d0] = 80000000h
+	        e5 93 f0 08   ldr     pc=>LAB_c00fff80,[r3,#offset ->SP]
+
+	    But slightly different in early games:
+
+	        e1 2f ff 1e   bx      lr
+
+	    Since this code has a predictable byte signature, we can search
+	    in memory to find its exact start address, then consume enough cycles to
+	    reduce the number of instructions executed until the next IRQ is asserted.
+	*/
+	if (m_irq_wait_start_addr == UNKNOWN_ADDR) {
+		if (m_maincpu->pc() > 0xc00cc000 && m_maincpu->pc() < 0xc00fff80) {
+			int8_t addr_delta = 8;
+			memory_share *shared = machine().root_device().memshare("workram");
+			uint32_t *shared32 = reinterpret_cast<uint32_t *>(shared->ptr());
```

Comment:
Please use a `required_shared_ptr<uint32_t>` for this to avoid the tag map lookup and help validation.

Reply:
Seems like I already had `m_workram` for this, so I just reused it here.